File: src/handlers/v2/tasks.go
package v2

import (
	"net/http"

	"aegis/consts"
	"aegis/database"
	"aegis/dto"
	"aegis/handlers"
	"aegis/repository"
	producer "aegis/service/producer"
	"aegis/utils"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

var wsUpgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 4096,
	CheckOrigin: func(r *http.Request) bool {
		return true // Allow all origins (JWT already handles auth)
	},
}

// BatchDeleteTasks
//
//	@Summary		Batch delete tasks
//	@Description	Batch delete tasks by IDs
//	@Tags			Tasks
//	@ID				batch_delete_tasks
//	@Accept			json
//	@Produce		json
//	@Security		BearerAuth
//	@Param			batch_delete	body		dto.BatchDeleteTaskReq		true	"Batch delete request"
//	@Success		200				{object}	dto.GenericResponse[any]	"Tasks deleted successfully"
//	@Failure		400				{object}	dto.GenericResponse[any]	"Invalid request format or parameters"
//	@Failure		401				{object}	dto.GenericResponse[any]	"Authentication required"
//	@Failure		403				{object}	dto.GenericResponse[any]	"Permission denied"
//	@Failure		500				{object}	dto.GenericResponse[any]	"Internal server error"
//	@Router			/api/v2/tasks/batch-delete [post]
func BatchDeleteTasks(c *gin.Context) {
	var req dto.BatchDeleteTaskReq
	if err := c.ShouldBindJSON(&req); err != nil {
		dto.ErrorResponse(c, http.StatusBadRequest, "Invalid request format: "+err.Error())
		return
	}

	if err := req.Validate(); err != nil {
		dto.ErrorResponse(c, http.StatusBadRequest, "Invalid request parameters"+err.Error())
		return
	}

	err := producer.BatchDeleteTasks(req.IDs)
	if handlers.HandleServiceError(c, err) {
		return
	}

	dto.JSONResponse[any](c, http.StatusNoContent, "Tasks deleted successfully", nil)
}

// GetTask handles getting a single task by ID
//
//	@Summary		Get task by ID
//	@Description	Get detailed information about a specific task
//	@Tags			Tasks
//	@ID				get_task_by_id
//	@Produce		json
//	@Security		BearerAuth
//	@Param			task_id	path		string									true	"Task ID"
//	@Success		200		{object}	dto.GenericResponse[dto.TaskDetailResp]	"Task retrieved successfully"
//	@Failure		400		{object}	dto.GenericResponse[any]				"Invalid task ID"
//	@Failure		401		{object}	dto.GenericResponse[any]				"Authentication required"
//	@Failure		403		{object}	dto.GenericResponse[any]				"Permission denied"
//	@Failure		404		{object}	dto.GenericResponse[any]				"Task not found"
//	@Failure		500		{object}	dto.GenericResponse[any]				"Internal server error"
//	@Router			/api/v2/tasks/{task_id} [get]
//	@x-api-type		{"sdk":"true"}
func GetTask(c *gin.Context) {
	taskID := c.Param(consts.URLPathTaskID)
	if !utils.IsValidUUID(taskID) {
		dto.ErrorResponse(c, http.StatusBadRequest, "Invalid task ID")
		return
	}

	resp, err := producer.GetTaskDetail(taskID)
	if handlers.HandleServiceError(c, err) {
		return
	}

	dto.SuccessResponse(c, resp)
}

// ListTasks handles simple task listing
//
//	@Summary		List tasks
//	@Description	Get a simple list of tasks with basic filtering via query parameters
//	@Tags			Tasks
//	@ID				list_tasks
//	@Produce		json
//	@Security		BearerAuth
//	@Param			page		query		int												false	"Page number"	default(1)
//	@Param			size		query		int												false	"Page size"		default(20)
//	@Param			task_type	query		consts.TaskType									false	"Filter by task type"
//	@Param			immediate	query		bool											false	"Filter by immediate execution"
//	@Param			trace_id	query		string											false	"Filter by trace ID (uuid format)"
//	@Param			group_id	query		string											false	"Filter by group ID (uuid format)"
//	@Param			project_id	query		int												false	"Filter by project ID"
//	@Param			state		query		consts.TaskState								false	"Filter by state"
//	@Param			status		query		consts.StatusType								false	"Filter by status"
//	@Success		200			{object}	dto.GenericResponse[dto.ListResp[dto.TaskResp]]	"Tasks retrieved successfully"
//	@Failure		400			{object}	dto.GenericResponse[any]						"Invalid request format or parameters"
//	@Failure		401			{object}	dto.GenericResponse[any]						"Authentication required"
//	@Failure		403			{object}	dto.GenericResponse[any]						"Permission denied"
//	@Failure		500			{object}	dto.GenericResponse[any]						"Internal server error"
//	@Router			/api/v2/tasks [get]
//	@x-api-type		{"sdk":"true"}
func ListTasks(c *gin.Context) {
	var req dto.ListTaskReq
	if err := c.ShouldBindQuery(&req); err != nil {
		dto.ErrorResponse(c, http.StatusBadRequest, "Invalid request format : "+err.Error())
		return
	}

	if err := req.Validate(); err != nil {
		dto.ErrorResponse(c, http.StatusBadRequest, "Invalid request parameters: "+err.Error())
		return
	}

	resp, err := producer.ListTasks(&req)
	if handlers.HandleServiceError(c, err) {
		return
	}

	dto.SuccessResponse(c, resp)
}

// GetTaskLogsWS handles WebSocket connections for real-time job log streaming
//
//	@Summary		Stream task logs via WebSocket
//	@Description	Establishes a WebSocket connection to stream real-time logs.
//	@Description	Process: 1. Validate Token -> 2. Push historical logs from Loki -> 3. Subscribe to Redis for real-time updates -> 4. Close on task completion.
//	@Tags			Tasks
//	@ID				get_task_logs_ws
//	@Param			task_id	path		string				true	"Task ID"
//	@Param			token	query		string				true	"JWT authentication token"
//	@Success		101		{object}	dto.WSLogMessage	"WebSocket connection established"
//	@Failure		400		"Invalid task ID"
//	@Failure		401		"Authentication failed"
//	@Failure		404		"Task not found"
//	@Router			/api/v2/tasks/{task_id}/logs/ws [get]
func GetTaskLogsWS(c *gin.Context) {
	taskID := c.Param(consts.URLPathTaskID)
	if taskID == "" {
		dto.ErrorResponse(c, http.StatusBadRequest, "task_id is required")
		return
	}

	// Authenticate via query parameter (WebSocket doesn't support custom headers)
	token := c.Query("token")
	if token == "" {
		dto.ErrorResponse(c, http.StatusUnauthorized, "token query parameter is required")
		return
	}

	if _, err := utils.ValidateToken(token); err != nil {
		dto.ErrorResponse(c, http.StatusUnauthorized, "invalid token: "+err.Error())
		return
	}

	// Verify task exists
	task, err := repository.GetTaskByID(database.DB, taskID)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			dto.ErrorResponse(c, http.StatusNotFound, "task not found")
			return
		}

		dto.ErrorResponse(c, http.StatusInternalServerError, "failed to retrieve task: "+err.Error())
		return
	}

	// Upgrade to WebSocket
	conn, err := wsUpgrader.Upgrade(c.Writer, c.Request, nil)
	if err != nil {
		logrus.Errorf("WebSocket upgrade failed for task %s: %v", taskID, err)
		return
	}
	defer conn.Close()

	// Delegate all streaming logic to the service layer
	streamer := producer.NewTaskLogStreamer(conn, taskID)
	streamer.StreamLogs(c.Request.Context(), task)
}

File: frontend/src/api/tasks.ts
/**
 * 任务 API
 * 使用 @rcabench/client SDK，手工实现缺失的端点
 */
import {
  TasksApi,
  TracesApi,
  type GroupStats,
  type ListTaskResp,
  type ListTasksTaskType,
  type StatusType,
  type TaskDetailResp,
  type TaskState,
} from '@rcabench/client';
import { apiClient, createApiConfig } from './config';

export const taskApi = {
  // ==================== SDK 方法 ====================

  /**
   * 获取任务列表 - 使用 SDK
   */
  getTasks: async (params?: {
    page?: number;
    size?: number;
    taskType?: ListTasksTaskType;
    immediate?: boolean;
    traceId?: string;
    groupId?: string;
    projectId?: number;
    state?: TaskState;
    status?: StatusType;
  }): Promise<ListTaskResp | undefined> => {
    const api = new TasksApi(createApiConfig());
    const response = await api.listTasks({
      page: params?.page,
      size: params?.size,
      taskType: params?.taskType,
      immediate: params?.immediate,
      traceId: params?.traceId,
      groupId: params?.groupId,
      projectId: params?.projectId,
      state: params?.state,
      status: params?.status,
    });
    return response.data.data;
  },

  /**
   * 获取任务详情 - 使用 SDK
   */
  getTask: async (taskId: string): Promise<TaskDetailResp | undefined> => {
    const api = new TasksApi(createApiConfig());
    const response = await api.getTaskById({ taskId });
    return response.data.data;
  },

  /**
   * 获取追踪组统计 - 使用 SDK
   */
  getGroupStats: async (groupId: string): Promise<GroupStats | undefined> => {
    const api = new TracesApi(createApiConfig());
    const response = await api.getGroupStats({ groupId });
    return response.data.data;
  },

  // ==================== 手工实现 (SDK 缺失) ====================

  /**
   * 批量删除任务 - 手工实现 (SDK 缺失)
   */
  batchDelete: (ids: string[]) =>
    apiClient.post('/tasks/batch-delete', { ids }),
};

/**
 * 创建实时日志流 (SSE)
 * 注意：EventSource 不支持自定义 headers，
 * 如果需要认证，后端需要支持通过 query params 传递 token
 */
export const createLogStream = (traceId: string): EventSource => {
  const token = localStorage.getItem('access_token');
  // SSE 连接通常需要通过 URL 参数传递 token
  const url = `/api/v2/traces/${traceId}/stream${token ? `?token=${encodeURIComponent(token)}` : ''}`;

  return new EventSource(url);
};
